package.json

```diff
@@ -108,6 +108,7 @@
     "jest-extended": "^1.1.0",
     "nodemon": "^2.0.22",
     "request": "^2.88.2",
+    "sass": "^1.79.3",
```

Comment:
> converting all styled-components usage in those parts over to either regular CSS/SASS, or to CSS Modules, depending on the scope of the usage.

Could you write a little about the guidelines you used to decide which system to use, and also why SASS was needed?

Reply:
Generally, if a style or set of styles was specific to an individual component, I moved it to a CSS Module sidecar file; if there was styling that was shared across multiple components, or that was affecting HTML elements that were part of the basic layout, I put it into global.css. 

I'm also, as much as possible, removing rules that aren't having any effect (due to overriding rules higher in the cascade), to reduce the amount of surface that needs to be considered when understanding how the styling works. 

With regard to SASS, as we discussed yesterday, I reached for that because I didn't realize that [CSS Nesting](https://caniuse.com/?search=css+nesting) was now a mainstream browser feature. I'll take the SASS stuff back out and just do everything as plain 'ol CSS.

Reply:
> With regard to SASS, as we discussed yesterday, I reached for that because I didn't realize that [CSS Nesting](https://caniuse.com/?search=css+nesting) was now a mainstream browser feature. I'll take the SASS stuff back out and just do everything as plain 'ol CSS.

One issue that I've run into with switching to plain CSS in CSS Modules, is that CSS Modules requires selectors to be "pure", and doesn't support the nesting selector, so you can't do something like: 

```
.componentClassName {
  margin: 0 auto;

  & h1 {
    font-weight: 500;
  }

  & a { 
    color: black;
    text-decoration: none;
  }
}
```

you best you can do would be something like this: 

```
.componentClassName {
  margin: 0 auto;
}
.componentClassName > h1 { 
  font-weight: 500;
}
.componentClassName > a { 
  color: black;
  text-decoration: none;
}
```

which is …fairly annoying by comparison.

Reply:
> One issue that I've run into with switching to plain CSS in CSS Modules, is that CSS Modules requires selectors to be "pure", and doesn't support the nesting selector…

I'm confused by your example. I believe you could write it without the nesting selector as:

```css
.componentClassName {
  margin: 0 auto;

  h1 {
    font-weight: 500;
  }

  a { 
    color: black;
    text-decoration: none;
  }
}
```

and have the same effect. An older version of the spec required `&` in this case, but that's no longer the case. I don't know why CSS Modules would support one and not the other at this point.

Stepping back though, do we think CSS nesting is widely supported enough yet? It was only supported across all browsers as of December 2023. The usage and date relative views on the [Can I use… page](https://caniuse.com/css-nesting) for it show a small but maybe significant number of older browser usage lacking support. (Our actual visitor usage may vary from Can I use…'s numbers.)

Reply:
> I'm confused by your example. I believe you could write it without the nesting selector as:
> 
> ```css
> .componentClassName {
>   margin: 0 auto;
[snip]

This code:

```css
.teamPage {
  h1 {
    color: var(--darkGrey);
  }
}
```

produces this output:

```
Syntax error: Selector "h1" is not pure (pure selectors must contain at least one local class or id)

  1 | .teamPage {
> 2 |   h1 {
    |   ^
  3 |     color: var(--darkGrey);
```

> Stepping back though, do we think CSS nesting is widely supported enough yet? It was only supported across all browsers as of December 2023. The usage and date relative views on the [Can I use… page](https://caniuse.com/css-nesting) for it show a small but maybe significant number of older browser usage lacking support. (Our actual visitor usage may vary from Can I use…'s numbers.)

I wondered about this too but didn't go beyond wondering, as I wasn't 100% sure what our browser support policy was.

The other potentially significant thing, which I've already run into once, is that SASS/SCSS supports [`@extend`](https://sass-lang.com/documentation/at-rules/extend/) and I don't think there's any equivalent for that in standard CSS. There are a number of `styled-elements` rules that are defined as:

```js
const NewRule = styled(OtherRule)`selector: rule`;
```
Porting those to a future new state is _way_ easier with `@extend`.

Reply:
> This code […] produces this output […]

Well that's dumb! I misread your prior statement as «CSS Modules require selectors to be "pure" **and so** don't support the nesting selector (`&`)» and so I thought "Ah, but the nesting selector isn't required for that example." But the correct reading of your statement is actually, «CSS Modules require selectors to be "pure" **and also** don't support the nesting selector (`&`)». Thanks for the clarifying example.

Reply:
> Well that's dumb!

I was less than impressed to discover it -- it apparently has to be that way in order to prevent "impure" selectors escaping the module scope and affecting the global document, from which I infer that the underlying implementation is, well, bad. 

While I'm bitching, it's also pretty underwhelming that using the CSS Modules still requires you to set a `className` prop to something imported from the CSS file, instead of just applying the CSS module to the entire hunk (and only the hunk) of JSX coming out of the component.

Reply:
> it apparently has to be that way

"has", sure. Ugh. If this works

```css
.componentClassName {
}
.componentClassName h1 { 
}
```

but this doesn't

```css
.componentClassName {
  h1 {
  }
}
```

it's extra dumb because they're equivalent. Difference is purely syntactical, so the issue is with the logic behind CSS Modules "pure" mode, not with anything fundamental. (To be clear, not complaining about you/your words, but about this dumb thing.)



> …instead of just applying the CSS module to the entire hunk (and only the hunk) of JSX coming out of the component.

I think you get that by doing:

```js
import "./foo.css";
```

in the `app/…/page.tsx`. That is, not using CSS Modules at all.

Reply:
> > …instead of just applying the CSS module to the entire hunk (and only the hunk) of JSX coming out of the component.
> 
> I think you get that by doing:
> 
> ```js
> import "./foo.css";
> ```
> 
> in the `app/…/page.tsx`. That is, not using CSS Modules at all.

I don't think that's quite what I want either -- as I understand it, doing `import "foo.css"` adds stuff to the stylesheet for the entire document; it's not scoped to just the component. So a top level `h1` rule in `foo.css` would apply to an `<h1>` that was in the document but didn't arise from component doing the import.

Reply:
Ahh, you're totally right.